src/bio/bio_device.c

```diff
@@ -780,109 +762,199 @@ led_device_action(void *ctx, struct spdk_pci_device *pci_device)
+		break;
+	default:
+		D_ERROR("invalid action supplied: %d\n", action);
+		return -DER_INVAL;
 	}
 
-	if (led_state_str == NULL)
-		return -DER_INVAL;
+	spdk_pci_for_each_device(&opts, led_device_action);
 
-	/* Determine SPDK LED state based on led_state string */
-	for (led_state = SPDK_VMD_LED_STATE_OFF;
-	     led_state <= SPDK_VMD_LED_STATE_REBUILD;
-	     led_state++) {
-		if (strcmp(led_state_str, g_led_states[led_state]) == 0) {
-			opts.led_state = (enum spdk_vmd_led_state)led_state;
-			break;
-		}
+	if (opts.status != 0) {
+		D_ERROR("LED %s failed (targeted state: %s): %s\n", LED_ACTION_NAME(action),
+			LED_STATE_NAME(*state), spdk_strerror(opts.status));
+		return opts.status;
+	}
+	if (!opts.all_devices && !opts.finished) {
+		D_ERROR("Device could not be found\n");
+		return -DER_NONEXIST;
 	}
 
-skip_led_str:
-	if (opts.led_state == SPDK_VMD_LED_STATE_UNKNOWN) {
-		D_ERROR("LED state is not valid or supported\n");
+	*state = opts.led_state;
+
+	return 0;
+}
+
+static int
+dev_uuid2pci_addr(struct spdk_pci_addr *pci_addr, uuid_t dev_uuid)
+{
+	struct bio_bdev		*bio_dev;
+	struct bio_dev_info	 b_info = { 0 };
+	int			 rc = 0;
+
+	if (pci_addr == NULL)
 		return -DER_INVAL;
+
+	bio_dev = lookup_dev_by_id(dev_uuid);
+	if (bio_dev == NULL) {
+		D_ERROR("Failed to find dev "DF_UUID"\n", DP_UUID(dev_uuid));
+		return -DER_NONEXIST;
 	}
 
 	rc = fill_in_traddr(&b_info, bio_dev->bb_name);
-	if (rc != 0) {
-		D_ERROR("Unable to get traddr for device %s\n", bio_dev->bb_name);
+	if (rc) {
+		D_ERROR("Unable to get traddr for device:%s\n", bio_dev->bb_name);
 		return -DER_INVAL;
 	}
 
-	rc = spdk_pci_addr_parse(&opts.pci_addr, b_info.bdi_traddr);
+	rc = spdk_pci_addr_parse(pci_addr, b_info.bdi_traddr);
 	if (rc != 0) {
 		D_ERROR("Unable to parse PCI address for device %s (%s)\n", b_info.bdi_traddr,
 			spdk_strerror(-rc));
-		D_GOTO(free_traddr, rc = -DER_INVAL);
+		rc = -DER_INVAL;
 	}
 
-	spdk_pci_for_each_device(&opts, led_device_action);
+	D_FREE(b_info.bdi_traddr);
+	return rc;
+}
 
-	if (opts.status != 0) {
-		D_ERROR("Setting LED state failed\n");
-		D_GOTO(free_traddr, rc = opts.status);
-	}
-	if (!opts.all_devices && !opts.finished) {
-		D_ERROR("Device with traddr %s could not be found\n", b_info.bdi_traddr);
-		D_GOTO(free_traddr, rc = -DER_NONEXIST);
-	}
+int
+bio_led_manage(struct bio_xs_context *xs_ctxt, char *tr_addr, uuid_t dev_uuid, unsigned int action,
+	       unsigned int *state)
+{
+	struct spdk_pci_addr	pci_addr;
+	int			rc;
 
-	/* Update current LED state after action */
-	bio_dev->bb_led_state = opts.led_state;
+	/**
+	 * If tr_addr is already provided, convert to a PCI address. If tr_addr is NULL or empty,
+	 * derive PCI address from the provided UUID and if tr_addr is an empty string buffer then
+	 * populate with the derived address.
+	 */
 
-	if (!reset && (opts.led_state != SPDK_VMD_LED_STATE_OFF)) {
-		/* Init the start time for the LED for a new event. */
-		bio_dev->bb_led_start_time = d_timeus_secdiff(0);
```

Comment:
This was for tracking the LED identify (blinking) start time, see bio_led_event_monitor(). I think that's why you observed blinking doesn't get reset anymore?

Reply:
I'm addressing this in the follow-up PR for configurable timeout.

Reply:
addressed in https://github.com/daos-stack/daos/pull/10768